### PR TITLE
feat(profiling): properly track Event Loop with `uvloop`

### PR DESCRIPTION
**Pull request by KowalskiThomas**
## Description

This PR adds support for profiling applications using `uvloop` with `asyncio`.

image

**Why is this needed?**

The `uvloop` Event Loop Policy doesn't inherit from `BaseDefaultEventLoopPolicy`, and on Python 3.11+ `uvloop.run` uses `asyncio.Runner` which bypasses `set_event_loop` entirely. This means the existing hooks for tracking event loops don't work with `uvloop`.  

Additionally, due to the internals not being the same, the logic we have today for stacking Task Stacks while excluding repeated parts of them doesn't yield good results with `uvloop`, which requires additional logic.

**How does it work?**

We patch `uvloop` functions (`uvloop.new_event_loop` for example) the same way we patch `asyncio` functions to detect e.g. when a new Task is created and to infer Task relationship using the existing `stack.track_asyncio_loop` and links the loop via `THREAD_LINK.link_object`, matching the behaviour of the existing `asyncio` event loop tracking.

An important thing to keep in mind is that, while it's pretty rare, it is not illegal to have more than one Thread with a running Loop. On top of that, nothing prevents developers from using `uvloop` in certain Threads and using the stock `asyncio` Event Loop in others –  this means we need to make sure we keep track of whether `uvloop` is in use on a per-Thread basis (not needed for Task relationships, but absolutely necessary for correctly unwinding/creating nice Task Stacks).  

Another important thing to keep in mind when tracking this is that there are several ways to use `uvloop`:
- `uvloop.install(); asyncio.run(coro())` which is deprecated but still available
- `uvloop.run(coro())` directly which is the recommended way to use `uvloop`

... we need to make sure we support both.

## Testing

Manual testing with a `uvloop`-based application, including Rapid APIs in dd-source 🚀 

I added support for `uvloop` environments for each Python version in the `riotfile` and updated one `asyncio` test to test against `uvloop` as well to validate the approach.   
I also plan to update all `asyncio`-related tests in `dd-trace-py` to also test against `uvloop`. In order to keep the current PR not too big, I will take care of it in https://github.com/DataDog/dd-trace-py/pull/16166/s.

On top of that, I added unit tests for the edge cases I mentioned with Threads, and some tests that check we don't accidentally assume `uvloop` is in use when it isn't (e.g. `uvloop` is imported but not actually used).

## Risks

Given that all existing tests pass, I think the worst case scenario would be that we have missed an edge case within the support for `uvloop`; I don't believe this PR can break existing features.

### CODE REVIEW

File: ddtrace/profiling/_asyncio.py [hunk 1]
```diff
@@ -256,3 +256,37 @@ def get_event_loop_for_thread(thread_id: int) -> typing.Optional["asyncio.Abstra
...
+    """Hook uvloop to track event loops.
+
+    uvloop doesn't inherit from BaseDefaultEventLoopPolicy, and on Python 3.11+
+    uvloop.run() uses asyncio.Runner which bypasses set_event_loop entirely.
+    We hook new_event_loop to catch all uvloop loop creations.
+    """
+    global THREAD_LINK
+
+    if THREAD_LINK is None:
+        THREAD_LINK = _threading._ThreadLink()
```
Comment (P403n1x87) [inline ddtrace/profiling/_asyncio.py]: Can we DRY this check?

File: ddtrace/profiling/_asyncio.py
Comment (taegyunkim) [inline ddtrace/profiling/_asyncio.py:1]: Do you mind adding a regression test if that's not too hard to do so?
We can add a new riot venv with uvloop and try re-using some of the existing asyncio tests you've added.

Comment (KowalskiThomas) [reply ddtrace/profiling/_asyncio.py:1]: I was planning to, but this PR is far from ready for review; it's still very much a WIP

File: ddtrace/profiling/_asyncio.py (hunk 1 above)
Comment (KowalskiThomas) [reply ddtrace/profiling/_asyncio.py]: Noted. The PR isn't really ready for review anyway at this point 😅

File: ddtrace/profiling/_asyncio.py
Comment (KowalskiThomas) [reply ddtrace/profiling/_asyncio.py:1]: I still don't consider the PR as ready (need to clean some things up) but I've already added a new `riot` environment for each Python version which will run `asyncio` tests using `uvloop.run` rather than `asyncio.run`. (For the time being, it only runs two tests, but I will extend it.)

File: ddtrace/profiling/_asyncio.py (hunk 1 above)
Comment (KowalskiThomas) [reply ddtrace/profiling/_asyncio.py]: Upon further scrutiny, I'm not sure it's the best thing to do. If I'm not mistaken, we only have two occurrences of that check, so it's not super bad. If I change it to e.g. a function, I will probably need to add an `assert` after the function call to be type safe, so it would still be as much code as before. 

Alternatively, I can change it to `THREAD_LINK = THREAD_LINK or _threading._ThreadLink()` but that's not necessarily much better.

What do you think?

File: ddtrace/internal/datadog/profiling/stack/src/echion/tasks.cc [hunk 2]
```diff
@@ -129,6 +129,32 @@ TaskInfo::create(TaskObj* task_addr)
       reinterpret_cast<PyObject*>(task_addr), loop, std::move(*maybe_coro), name, std::move(waiter));
 }
 
+// When uvloop.run() is used, the top-level Task contains a wrapper coroutine
+// named "run.<locals>.wrapper" that just validates the loop type and awaits the user's main
+// coroutine. We skip this frame to keep the stack clean and consistent with regular asyncio.
+bool
+is_uvloop_wrapper_frame(const Frame& frame)
```
Comment (P403n1x87) [inline ddtrace/internal/datadog/profiling/stack/src/echion/tasks.cc]: I wonder if we could design this in a way where we have some sort of compat layer that can (post?)process data depending on what framework is in use. I would like if we could avoid adding product-specific logic in code units that have a different concern

Comment (KowalskiThomas) [reply ddtrace/internal/datadog/profiling/stack/src/echion/tasks.cc]: Overall I do agree, I'm just slightly concerned about what this API would look like, and how we'd handle certain things like the special `run` wrapper coroutine that `uvloop` uses whenever you use `uvloop.run` –  this sounds like it needs to be handled separately.

Alternatively, we could also post-process all that somewhere else (either locally, or maybe even in the profiling backend) but then I'd be concerned that:
1. We lose important information, e.g. when we stack Stacks, before we stack them, we can strip certain top or bottom parts of them; if we need to do that after they've been stacked it's more complicated and error-prone
2. It would make testing more of a hassle as the Profiles we'd be sending wouldn't be the Profiles users would see; ideally we should be able to test that the Profiles reported by the library are (at least almost) the same as the ones users see

**Comment by KowalskiThomas** [comment]
> @taegyunkim What would happen if a user imports both uvloop and asyncio and but doesn't use uvloop's loop?
> Would we still be able to show the tasks from asyncio loop?

We call `stack.set_uvloop_mode(True)` as soon as the user imports `uvloop` indeed, so I do believe if they import `uvloop` we will do a worse job at detecting (and splitting at) "boundary frames" like I call them in the code.

I'm not sure how I can make that better –  potentially calling this when the user calls `uvloop.run`? What do you think?

**Comment by KowalskiThomas** [comment]
Notes from @taegyunkim:
- Let's set `uvloop` mode in `uvloop.run` instead of when we import `uvloop`
  -  We also need to handle  `uvloop.install()` then `asyncio.run(whatever())`
- What if we have some Threads with `uvloop` and some threads without (and add a test)
  - We need to have a "use `uvloop`" flag per Thread.

---

EDIT: there's a slightly more complicated case: `uvloop.install` (which is deprecated, but that we can't expect people never to use) doesn't install _for the current Thread_, it installs the special `uvloop` for all Threads, which means:
- If another Thread hasn't started its Task yet, then it'll use  `uvloop` when it does
- If another Thread has already started its Task, then it won't use `uvloop` for the current Task, but will in the future... 

This is very racy by nature and I'm not sure how to handle it. We could do something clever to cover all cases but I fear we may over engineer this, as having (1) several Threads running `asyncio` Tasks AND (2) one of them, not the main one, calling `uvloop.install` AND (3)  not using `uvloop.run` is a very edgy edge case... 

The following code allows to show the race / make it deterministically take a path or the other.

```py
import asyncio
import math
from threading import Event
from threading import Thread
import time

import uvloop


condition = Event()


async def async_main(parent: str) -> int:
    print("in main from ", parent)
    print(asyncio.get_running_loop())
    result = 2
    start_time = time.time()
    while time.time() - start_time < 1:
        result *= math.factorial(min(result, 10000))

    return result


def other_thread():
    print("other_thread")

    print("other_thread: waiting for condition")
    # condition.wait()
    print("other_thread: condition notified")

    # uvloop.run(async_main("other_thread"))

    uvloop.install()
    condition.set()
    asyncio.run(async_main("other_thread"))

    print("other_thread: done")


def main():
    async def async_main_wrapper():
        print("in main")
        # condition.set()
        print("main: condition notified")
        return await async_main("main")

    t = Thread(target=other_thread)
    t.start()
    condition.wait()
    asyncio.run(async_main_wrapper())
    t.join()


main()

```

**Comment by KowalskiThomas** [comment]
I just rebased since I had changes on main that conflicted for removal of global/static state.

File: ddtrace/internal/datadog/profiling/stack/src/echion/tasks.cc (hunk 2 above)
Comment (KowalskiThomas) [reply ddtrace/internal/datadog/profiling/stack/src/echion/tasks.cc]: Fixed!

**Comment by r1viollet** [comment]
Do we need to add a setting to remove the behaviour if we have an issue while deploying this to customers ?

**Comment by KowalskiThomas** [comment]
>  Do we need to add a setting to remove the behaviour if we have an issue while deploying this to customers ?

Great point, I didn't even realise this was possible! I can add this as a configuration entry/setting so that customers can disable it if needed.

---

EDIT: I added a configuration that can be flipped with `DD_PROFILING_STACK_UVLOOP` (default behaviour is "enable uvloop sampling").

**Comment by r1viollet** [comment]
I'll give this a spin to check overhead, though overall this looks good. Thanks for thinking about tests.

**Comment by KowalskiThomas** [comment]
> @r1viollet  I'll give this a spin to check overhead, though overall this looks good. 

Note that the overhead should be good _compared to stock `asyncio`_, but is probably pretty bad compared to non-async Profiling. It's a known issue that we have some ideas to work around, but yeah, don't expect miracles there (yet!)

File: ddtrace/internal/settings/profiling.py [hunk 3]
```diff
@@ -269,6 +269,14 @@ class ProfilingConfigStack(DDConfig):
         private=True,
     )
 
+    uvloop = DDConfig.v(
```
Comment (taegyunkim) [inline ddtrace/internal/settings/profiling.py:272]: Adding a new conf requires some changes in dd-go following 

> E                   ValueError: (NOT A FLAKE) Read this quick runbook to update allowed configs: https://github.com/DataDog/system-tests/blob/main/docs/edit/runbook.md#test_config_telemetry_completeness

As noted in https://github.com/DataDog/dd-trace-py/actions/runs/21678944718/job/62509117992?pr=16100

Comment (KowalskiThomas) [reply ddtrace/internal/settings/profiling.py:272]: Ah, I saw _not a flake_ and I thought... it was a flake. Smart of me.

Comment (KowalskiThomas) [reply ddtrace/internal/settings/profiling.py:272]: Done here: https://github.com/DataDog/dd-go/pull/220862

**Comment by KowalskiThomas** [comment]
**Note** we're currently looking at this in DoE to validate overhead, but we can't seem interesting stuff for some reason. Will keep investigating there and hold off to merging until we have an answer.

---

After checking more, it seems like we are seeing the expected Frames / Stacks, including in DoE. Example screenshot for the following code.

```py
async def thomas():
    start = time.time()
    result = 2
    while time.time() - start < 3:
        result * math.factorial(result)

async def load() -> PlainTextResponse:
    if config.off_cpu > 0:
        await asyncio.sleep(config.off_cpu)
    if config.loops_num > 0:
        wrap(config.stack_depth, config.recursive, run_cpu_on)
    if config.exceptions:
        wrap(config.stack_depth, config.recursive, run_exception)

    await asyncio.create_task(thomas(), name="thomas")

    return PlainTextResponse("Hello World\n")
```

image


→  We see two stacked Frames, which is only possible if we properly track and unwind `uvloop`, as they are two separate Tasks.

File: ddtrace/profiling/_asyncio.py (hunk 1 above)
Comment (KowalskiThomas) [reply ddtrace/profiling/_asyncio.py]: Ended up DRYing it anyway :)

**Review by gnufede — APPROVED** [review]
